### ISSUE TITLE
cpu-set not work in openmpi4.0.5

### ISSUE THREAD

**Issue by oleotiger** [open-mpi/ompi#8669]
## Background information
I wanna to run wrf.exe with openmpi and bind processes to a centain set of cores. With cpu-set it doesn't work as it supposes to be.

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
openmpi v4.0.5

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
From a source

### Please describe the system on which you are running

* Operating system/version: CentOS Linux release 7.6.1810 (Core)
* Computer hardware:  24cores per socket, 2 sockets per node. 4 node  totally.  CPU: Intel(R) Xeon(R) 6248R
* Network type:  IB

-----------------------------

## Details of the problem

I have 4 hosts to run an openmpi application `wrf.exe`. There are 48 cores(24 per socket,2 socket per node) on each host.

I wanna run 40 processes to handle the application. Each node will run 10 processes on 10 cores (0,2,4,6,8,10,12,14,16,18).

Here are what I tried but I didn't get the mpi manner as I suppose  to be.

1. `mpirun --hostfile=4hosts --cpu-set 0,2,4,6,8,10,12,14,16,18 ...`. 10 cores run at 100% util of each node. But over 100 of processes `wrf.exe` run on each node.

2. `mpirun --hostfile=4hosts -np 40 --cpu-set 0,2,4,6,8,10,12,14,16,18 ...`. 40 processes of wrf.exe runs on the first host in 4hosts file. Only 10 cores on the host run at 100% util.

3. 
`mpirun --hostfile=4hosts -np 40 -N 10 --cpu-set 0,2,4,6,8,10,12,14,16,18 ...` 

 `mpirun --hostfile=4hosts -N 10 --cpu-set 0,2,4,6,8,10,12,14,16,18 ...` 

 `mpirun --hostfile=4hosts ---map-by ppr:10:node  --cpu-set 0,2,4,6,8,10,12,14,16,18 ...` 

All got errors:
```
Conflicting directives for mapping policy are causing the policy
to be redefined:

  New policy:   RANK_FILE
  Prior policy:  UNKNOWN

Please check that only one policy is defined.
```

There are four ip in my hostfile which indicates the hosts I would like to run my application. I'm using openmpi on centos7.6.

How to use -cpu-set in mpirun to run total 40 process on 4 node, 10 process on specified cores on each node?

**Comment by rhc54** [comment]
It appears cpu-set is not working in the 4.x series. It is working, however, in the master and v5.0.x branches, so it should be released in the not-too-distant future. Meantime, you might want to try one of those two sources as I very much doubt the fix will be backported to the 4.x series.

**Comment by oleotiger** [comment]
From the second command as shown above, I found a way to run with cpu-set. It seems to act like the behavior as I suppose it to do but I don't know if it's correct.

Command: `mpirun --hostfile=4hosts -np 40 --cpu-set 0,2,4,6,8,10,12,14,16,18 .... `

In file 4hosts, I add slots=xxx.
```
node1 slots=10
node1 slots=10
node1 slots=10
```

**Comment by rhc54** [comment]
It is ignoring cpu-set - all it is doing is putting 10 procs on each node because you told it to do so with your "slots=10" argument. You might check to see where those procs are bound. I believe the default is bind-to socket. You might want to bind-to core.

**Comment by oleotiger** [comment]
The whole command is `mpirun --hostfile=4hosts -np 40 --cpu-set 0,2,4,6,8,10,12,14,16,18 --bind-to core  -x OMP_NUM_THREADS=1 -mca pml ucx -mca btl ^vader,tcp,openib,uct -mca io romio321 -x UCX_TLS=self,sm,rc -x UCX_NET_DEVICES=mlx5_0:1  -report-bindings    ./wrf.exe`

Among the four nodes, ranks on 3 nodes report:
```
[nodexxx:38845] MCW rank xxx is not bound (or bound to all available processors)
```
10 ranks on one node reports:
```
[nodexxx:08518] MCW rank xxx bound to socket 0[core 0[hwt 0]], socket 0[core 2[hwt 0]], socket 0[core 4[hwt 0]], socket 0[core 5[hwt 0]], socket 0[core 7[hwt 0]], socket 0[core 9[hwt 0]], socket 0[core 10[hwt 0]], socket 0[core 12[hwt 0]], socket 0[core 15[hwt 0]], socket 0[core 17[hwt 0]]: [B/./B/./B/B/./B/./B/B/./B/././B/./B/./././././.][./././././././././././././././././././././././.]
```
With command `htop`, processes are running on (start from 1) `1,3,5,7,8,11,13,15,16,20`. Not exact as cpu-set indicated.

**Comment by oleotiger** [comment]
There is a weird phenomenon.
If I start to application on 4 nodes:
`mpirun --hostfile=4hosts -np 40 --cpu-set 0-11,24-35 --bind-to core app_1`
and
`mpirun --hostfile=4hosts -np 40 --cpu-set 12-23,36-47 --bind-to core app_2`
at the same time, all cores are runing at 100% util.

As you mentioned above, mpi won't distribute ranks on the cores as cpu-set indicated ( may be random? as 10 ranks seems to distribute on 1,3,5,7,8,11,13,15,16,20). But 2 applications do occupy all the cores. If ranks are distrbuted randomly on sockets, I think there may be idle cores.

**Comment by oleotiger** [comment]
The purpose I wanna use `--cpu-set` is that: I wanna to run 2 applications at the same time and control their distribution on the core. For example, at one node, each application takes one socket or one numa node. And further, each application runs on half of cores of each numa node. I tried map-by numa, but both applications run on the first half cores of the numa node. 

Any ideas to achieve the purpose if cpu-set does not on version4.x?

**Comment by rhc54** [comment]
This has been fixed on OMPI's master branch which is soon to be released as v5.0. You are welcome to utilize a nightly tarball (https://www.open-mpi.org/nightly/master/ or https://www.open-mpi.org/nightly/v5.0.x/) in the interim.
